### ISSUE TITLE
consider to use FEATURES="nostrip" on some packages

### ISSUE THREAD

**Issue by pchome** [InBetweenNames/gentooLTO#9]
```
$ cat /etc/portage/env/nostrip 
FEATURES="${FEATURES} nostrip"
```
```
$ cat /etc/portage/package.env/lto 
sys-libs/zlib nostrip
```

`# emerge -1 sys-libs/zlib -va`
w/ freshly installed gcc 7.2.0 and new CFLAGS from [make.conf.lto](https://github.com/InBetweenNames/gentooLTO/blob/8eb65a8cd14ec2812d2ac99f63dd81366cd5e61d/sys-config/ltoize/files/make.conf.lto) (nothing else from overlay, just playing w/ optimizations for now) this command triggered rebuild for ~15 packages including `sys-apps/pciutils` which marked in overlay _workarounds_ as **nolto**

and everything **compiled** w/ default LTO flags

not sure about python which marked as **ltonographite** and potentially can break portage, but I can confirm problem w/  **net-misc/openssh** (lto/nolto.conf #hangs on exit with lto)
both are triggered by zlib too

I can suggest command (picked from one of GCC Upgrade manuals long time ago) to detect other potential candidates for this env: `# emerge -1 $(find /lib* /usr/lib* -type f -name "*.a") -vtp`

**Comment by pchome** [comment]
so far I checked some packages from nolto list (build and basic functions):
 - sys-apps/grep - ok
 - sys-apps/pciutils - ok
 - app-text/mupdf - ok
 - dev-libs/elfutils - KO
 - sys-libs/glibc - ok, but looks like it uses own optimisations, only -O2 and basic flags were detected in output
 - media-sound/pulseaudio - KO, something w/ libltdl and dlpreopened or so, similar to https://bugs.freedesktop.org/show_bug.cgi?id=99675

and  ... wine w/ custom-cflags set  - KO ;-), pthread linking errors, maybe sys-libs/glibc somehow can be triggered to built w/ LTO and nostrip

**Comment by InBetweenNames** [comment]
Looking into this--thanks for your work!

**Comment by darkbasic** [comment]
Making sys-libs/glibc work with LTO would be awesome :D

**Comment by InBetweenNames** [comment]
Indeed!  @pchome I just tested building `grep`, `pciutils`, and `mupdf` with LTO and it worked.  Awesome!  I'm still looking into the `FEATURES` stuff, but I was wondering if you would mind setting up a separate PR just to get those things off the `ltoworkarounds.conf` list?  Perhaps by commenting them out.  I would do it myself but I think you deserve the credit on this one :)

**Comment by pchome** [comment]
feel free to reject PR
it was created mainly to point what must be done to fix some problems

**Comment by InBetweenNames** [comment]
OK no problem -- glibc 2.26 does indeed seem to "force" nolto by stripping all flags, including LTO ones, and replacing them with -O2.

~~~
 * strip-flags: CFLAGS: changed '-march=haswell -O3 -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=8 -fuse-linker-plugin -pipe' to '-march=haswell -pipe -O2'
 * strip-flags: CXXFLAGS: changed '-march=haswell -O3 -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=8 -fuse-linker-plugin -pipe' to '-march=haswell -pipe -O2'
 * strip-flags: LDFLAGS: changed '-Wl,--hash-style=gnu -march=haswell -O3 -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=8 -fuse-linker-plugin -pipe' to '-Wl,--hash-style=gnu -march=haswell -pipe -O2'
~~~

I modified the `toolchain-glibc` eclass to prevent that from happening, but it appears that glibc itself does not play nice with LTO just yet.  I'm going to leave the override in just so this is documented.

**Comment by InBetweenNames** [comment]
@pchome Just tested the FEATURES="nostrip" stuff.  On my system at least, sshd still hangs on session exit with LTO.  It doesn't with nolto however.  On your system, did you observe this behaviour too?

**Comment by pchome** [comment]
> I can confirm problem w/ net-misc/openssh (lto/nolto.conf #hangs on exit with lto)

yes

> I'm going to leave the override in just so this is documented.

maybe in another section than
w/ all other packages doing this and all `-custom-cflags` as default

is it possible somehow to point flags and versions in this configs? like `app-category/app-name::repo[flag]` ?
if so than definitely some `+custom-cflags`  packages mast be added to overrides
`app-emulation/wine-staging -lto -graphite` for example

or maybe  `custom-cflags`  flag masked
not sure, just FYI

**Comment by InBetweenNames** [comment]
I actually tried that before but it crashed portage--I don't think it was designed with that functionality in mind.  I was going to do it with perl originally, like:
~~~
*/*[perl] nolto.conf
~~~

**Comment by InBetweenNames** [comment]
@pchome Just curious, what effect does `FEATURES="nostrip"` have on your LTO?  Are you able to build any packages with it that would not otherwise be able to build?

**Comment by pchome** [comment]
`FEATURES="nostrip"` currently only set for `sys-libs/zlib` in my `env/`
and only tested w/ `sys-apps/pciutils` which fails to build w/o this trick, but you can recheck to be sure

then I "randomly" picked some packages from your `nolto` list and they ok w/ `lto` enabled
not sure if it zlib effect, or not full-lto system (I don't plan to re-emerge world) or something else

in addition to packages reported above:
sys-devel/gettext - ok w/ lto
dev-libs/libgcrypt - ok w/ lto

**Comment by InBetweenNames** [comment]
Interesting, because I managed to build `pciutils` just fine with zlib being `ltofat.conf`.  However I've not tested the other packages you've mentioned.  Probably a comb through the `nolto.conf` list is in order here.

**Comment by pchome** [comment]
Sorry for misunderstanding, but this issue about a goal of the project.
If the goal includes pure-lto than you can fix package and build it w/o fat objects.

Or you can use fat-lto, this will increase build time and size but no other fixing needed. I don't see other big differances in this methods.
So if fat-lto is ok than issue can be closed.

**Comment by pchome** [comment]
and this (about fat objects):
> A, perhaps surprising, side effect of this feature is that any mistake in the toolchain leads to LTO information not being used (e.g. an older libtool calling ld directly). This is both an advantage, as the system is more robust, and a disadvantage, as the user is not informed that the optimization has been disabled. 

https://gcc.gnu.org/onlinedocs/gccint/LTO-Overview.html

**Comment by InBetweenNames** [comment]
I can't reproduce being able to build without fat LTO objects, sorry :/ I have been able to knock off all packages you have mentioned on the `nolto` config however.  The reason `ltofat` is needed is in case a `nolto` package needs to link against an `lto` one, or the package itself has some LDFLAGS oddities inside it.  The final binaries are unaffected except the `.a` libraries, which will contain both LTO and non-LTO symbols with `ltofat.conf`.

Is this documentation up to date?  I'm finding some inconsistencies here versus what the Optimize-Options GCC page states.  Specifically, here it says it defaults to fat-lto-objects, but on the Optimize-Options it says it defaults to slim-lto-objects.  This looks more like a blueprint of what LTO would evolve into rather than being a description of the current implementation.

Since we use slim lto objects by default, `ld`, `nm`, `ar`, and `ranlib` called directly will attempt a non-LTO build, which will fail if any linked-to library isn't `nolto` or `ltofat.  So we can at least track these packages right now.  More is intended to be done here in the future.

The goal here isn't pure-lto just yet.  Right now I'd be happy if we could get everything `ltofat` at least.  As we move closer to our stated goals in the README, those goalposts will shift.

Thanks for your contributions in pointing out these packages which could have been removed from `nolto`.  That is super helpful and I have no doubt we will need more of that down the line as software improves.  I hope I'll be seeing more contributions from you in the future!
